### PR TITLE
Playback sidebar buttons

### PR DESCRIPTION
**Pull request by knguyen423**

### CODE REVIEW

**Review by claycrews2002 — CHANGES_REQUESTED** [review]
Screenshot 2023-03-26 at 10 37 07 AM

When clicking any of the buttons in the playback sidebar I get this error. My Spotify account isn't premium, which is my I'm getting the error. 
It doesn't look like player commands can be used by non-premium users. A different way to handle this would be good.

**Comment by cravend** [comment]
https://github.com/SCCapstone/DKMS/issues/251

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 1]
```diff
@@ -0,0 +1,148 @@
+"use client";
+
+import { useRouter } from "next/navigation";
+import { useState, useTransition } from "react";
+
+import getCurrentTrackUri from "@/lib/playback/gettrackrui";
+
+import play from "../../../../lib/playback/play";
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:8]: don't use dot imports, use `@/lib`

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import next from "../../../../lib/playback/skipnext";
+import prev from "../../../../lib/playback/skipprev";
+import BasePanel from "../BasePanel";
+
+let currentIsTrackPlaying = false;
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:13]: don't use global let it never works how you expect

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  isTrackPlaying,
+  uri,
+}: {
+  isTrackPlaying: boolean;
+  uri: string;
+}) => {
+  const router = useRouter();
+  const [isPending, startTransition] = useTransition();
+  const [isFetching, setIsFetching] = useState(false);
+  const [isPlaying, setIsPlaying] = useState(false); // new state variable for player state
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:25]: Shouldn't this default to `isTrackPlaying`?

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  };
+
+  const handleClick = async (e: React.MouseEvent) => {
+    e.preventDefault();
+    setIsFetching(true);
+    // TODO unsafe access on any value
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const tmp_uri = await getCurrentTrackUri();
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access
+    currentIsTrackPlaying = await play(tmp_uri.item.uri, currentIsTrackPlaying);
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:53]: What is this for?

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    e.preventDefault();
+    setIsFetching(true);
+    await next(uri, isTrackPlaying, true);
+    setIsFetching(false);
+    startTransition(() => {
+      router.refresh();
+    });
+  };
+
+  const handleClick = async (e: React.MouseEvent) => {
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:46]: this should be named `handlePlayPauseClick`

File: components/Sidebar/panels/Playback/Playback.tsx (hunk 4 above)
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:53]: Clicking play when a song is paused resets the song. You shouldn't pass in the URI if you just want to resume the current song, only if youre starting a new song

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                d="M11 6L4 12L11 18V6Z"
+                strokeWidth="2"
+                strokeLinecap="round"
+                strokeLinejoin="round"
+              />
+            </svg>
+          </button>
+          <button
+            type="button"
+            className="focus:outline-none"
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:116]: don't do this, it makes an app less accessible. You should probably have like className `btn btn-square btn-ghost` or maybe `btn-link`

File: components/Sidebar/panels/index.tsx [hunk 7]
```diff
@@ -1,12 +1,20 @@
...
-    <Friends />
-    {/* @ts-expect-error Next 13 handles async components */}
-    <Notifications />
-  </div>
-);
+const SidebarPanels = ({ isDesktop }: { isDesktop?: boolean }) => {
+  // Assume you have the following values for isTrackPlaying and uri
+  const isTrackPlaying = false;
+  console.log("SidebarPanels init");
+  const uri = "spotify:track:6y0igZArWVi6Iz0rj35c1Y";
```
Comment (cravend) [inline components/Sidebar/panels/index.tsx:9]: Why are you hard coding a URI?

File: components/Sidebar/panels/index.tsx [hunk 8]
```diff
@@ -1,12 +1,20 @@
...
-  <div className={isDesktop ? "hidden md:block" : "md:hidden"}>
-    <Friends />
-    {/* @ts-expect-error Next 13 handles async components */}
-    <Notifications />
-  </div>
-);
+const SidebarPanels = ({ isDesktop }: { isDesktop?: boolean }) => {
+  // Assume you have the following values for isTrackPlaying and uri
+  const isTrackPlaying = false;
+  console.log("SidebarPanels init");
```
Comment (cravend) [inline components/Sidebar/panels/index.tsx:8]: We shouldn't do console log

File: components/Sidebar/panels/Playback/Playback.tsx
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:1]: You should have an `index.tsx` server component getting the current player status and passing it into here. The SidebarPanels shouldn't directly import `Playback.tsx`, it should import the server component first.

File: components/Sidebar/panels/index.tsx [hunk 9]
```diff
@@ -1,12 +1,20 @@
 import Friends from "./Friends";
 import Notifications from "./Notifications";
+import Playback from "./Playback/Playback";
```
Comment (cravend) [inline components/Sidebar/panels/index.tsx:3]: Same as above, this should import an index.tsx file inside of playback (and that's where you could put isTrackPlaying and URI)

```suggestion
import Playback from "./Playback";
```

File: lib/playback/gettrackrui.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+type CurrentlyPlayingResponse = {
+  item: {
+    uri: string;
+  };
+};
+
+const getCurrentTrackUri = async () => {
+  const token = await getAccessToken();
+  const response = await fetch(
```
Comment (cravend) [inline lib/playback/gettrackrui.ts:11]: Don't fetch directly, use `getSpotifyData`

File: lib/playback/gettrackrui.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import getAccessToken from "../getAccessToken";
+
+type CurrentlyPlayingResponse = {
```
Comment (cravend) [inline lib/playback/gettrackrui.ts:3]: Don't manually type this, use the `SpotifyApi` namespace

File: lib/playback/gettrackrui.ts
Comment (cravend) [inline lib/playback/gettrackrui.ts:1]: this should be named `getCurrentTrackUri.ts` — name it with proper capitalization + match the default export for simplicity

File: lib/playback/pause.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+import getAccessToken from "../getAccessToken";
+
+const pause = async (uri: string, isTrackPlaying: boolean) => {
+  const method = isTrackPlaying ? "PUT" : "PUT";
+  const endpoint = isTrackPlaying ? "pause" : "play";
+  const url = `https://api.spotify.com/v1/me/player/pause${endpoint}`;
+  const response = await putSpotifyData(url, {
+    method,
+    headers: {
```
Comment (cravend) [inline lib/playback/pause.ts:11]: Don't pass in the header into `putSpotifyData` it does it automatically

File: lib/playback/play.ts
Comment (cravend) [inline lib/playback/play.ts:1]: What's the different between play and pause? This should probably be `togglePlaying` (which doesn't take in an URI) and `playTrack` (which does).

File: lib/playback/skipnext.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+import getAccessToken from "../getAccessToken";
+
+const next = async (
+  uri: string,
+  isTrackPlaying: boolean,
+  isSkipToPrevious = false
+) => {
+  let endpoint = "https://api.spotify.com/v1/me/player/play";
+  let method = "PUT";
```
Comment (cravend) [inline lib/playback/skipnext.ts:11]: don't use let, it's almost never needed. If you have to switch the method, you probably need to make a new function

File: lib/playback/skipnext.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  let body: string | null = JSON.stringify({
+    uris: [uri],
+  });
+
+  if (isSkipToPrevious) {
```
Comment (cravend) [inline lib/playback/skipnext.ts:16]: What does this do? Isn't this the goToNextSong function? Can you add JSDoc comments

File: lib/playback/skipnext.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  const headers = {
+    Authorization: JSON.stringify({
+      getAccessToken,
+    }),
+    "Content-Type": "application/json",
+  };
+
+  await putSpotifyData(endpoint, { method, headers, body });
```
Comment (cravend) [inline lib/playback/skipnext.ts:31]: You should return this (same for all the others)

File: lib/playback/skipnext.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+  if (isSkipToPrevious) {
+    endpoint = "https://api.spotify.com/v1/me/player/next";
+    method = "POST";
+    body = null;
+  } else if (isTrackPlaying) {
+    method = "DELETE";
+  }
+
+  const headers = {
```
Comment (cravend) [inline lib/playback/skipnext.ts:24]: No headers

File: lib/playback/skipprev.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import putSpotifyData from "@/lib/putSpotifyData";
+
+import getAccessToken from "../getAccessToken";
+
+const prev = async (
```
Comment (cravend) [inline lib/playback/skipprev.ts:5]: Same as `skipNext`

File: lib/playback/skipnext.ts
Comment (cravend) [inline lib/playback/skipnext.ts:1]: This should be called `skipNext.ts` with a capital and the function should also be called `skipNext`

**Review by cravend — CHANGES_REQUESTED** [review]

File: components/Sidebar/panels/Playback/Playback.tsx [hunk 18]
```diff
@@ -0,0 +1,148 @@
+"use client";
+
+import { useRouter } from "next/navigation";
+import { useState, useTransition } from "react";
+
+import getCurrentTrackUri from "@/lib/playback/gettrackrui";
+
+import play from "../../../../lib/playback/play";
+import next from "../../../../lib/playback/skipnext";
+import prev from "../../../../lib/playback/skipprev";
```
Comment (cravend) [inline components/Sidebar/panels/Playback/Playback.tsx:10]: Can you make an index file inside of playback and export all of these from there? That way, you can do `import { play, next, prev } from "@/lib/playback";` ?

**Comment by cravend** [comment]
https://daisyui.com/components/swap/ This might be useful for the play/pause

File: components/Sidebar/panels/Playback/Playback.tsx (hunk 4 above)
Comment (knguyen423) [reply components/Sidebar/panels/Playback/Playback.tsx]: Added since play was having issues knowing when the song was either playing or paused

**Pull request closed without merging**